### ISSUE TITLE
Why does this simple C++ XOR example not converge?

### ISSUE THREAD

**Issue by OleStauning** [Microsoft/CNTK#1629]
I am trying to figure out how to use the C++ from looking at the sparse documentation, header-files and code examples. I cannot figure out why the following simple example does not converge:

```
#include <iostream>
#include <vector>

#include "CNTKLibrary.h"

using namespace ::CNTK;

int main()
{
    std::vector< std::vector<float> > data_in = { {0,0}, {0,1}, {1,0}, {1,1} };
    std::vector< std::vector<float> > data_out = { {0}, {1}, {1}, {0} };

    auto device = DeviceDescriptor::CPUDevice();

    auto inputDim = 2UL;
    auto outputDim = 1UL;

    auto input = InputVariable({inputDim}, DataType::Float);
    auto truth = InputVariable({outputDim}, DataType::Float);

    Variable layer = input;
    auto hiddenWeights = Parameter(NDArrayView::RandomUniform<float>({ 8, 2 }, -0.05, 0.05, 1, device));
    auto hiddenBias = Parameter(NDArrayView::RandomUniform<float>({ 8 }, -0.05, 0.05, 1, device));
    auto hidden = ReLU(Plus(hiddenBias, Times(hiddenWeights, input)));
    auto outWeights = Parameter(NDArrayView::RandomUniform<float>({ 1, 8 }, -0.05, 0.05, 1, device));
    auto outBias = Parameter(NDArrayView::RandomUniform<float>({ 1 }, -0.05, 0.05, 1, device));
    auto prediction = Sigmoid(Plus(outBias, Times(outWeights, hidden)));
    auto trainingLoss = BinaryCrossEntropy(prediction, truth);

    auto schedule = LearningRatePerSampleSchedule({ .05 });
    auto momentum = MomentumAsTimeConstantSchedule({ 1.0, 10.0, 100.0 });
    auto learner = AdamLearner(prediction->Parameters(), schedule, momentum);
    auto trainer = CreateTrainer(prediction, trainingLoss, trainingLoss, { learner });

    const size_t minibatchSize = 10;

    std::vector< std::vector<float> > features(minibatchSize);
    std::vector< std::vector<float> > labels(minibatchSize);

    size_t numMinibatchesToTrain = 200;
    size_t outputFrequencyInMinibatches = 10;
    size_t trainIdx=0;
    for (size_t i = 0; i < numMinibatchesToTrain; ++i)
    {
        for(size_t j = 0; j< minibatchSize; ++j)
        {
            features[j] = data_in[trainIdx];
            labels[j] = data_out[trainIdx];
            trainIdx = (trainIdx + 1) % data_in.size();
        }
        auto batchInput = Value::Create({inputDim}, features, device);
        auto batchLabels = Value::Create({outputDim}, labels, device);
        std::unordered_map<Variable, ValuePtr> minibatchBindings = {
            { input, batchInput }, 
            { truth, batchLabels }
        };

        trainer->TrainMinibatch(minibatchBindings, device);
        if ((i%outputFrequencyInMinibatches) == 0)
            std::cout<<"Loss: "<<trainer->PreviousMinibatchLossAverage()<<std::endl;
    }
    
    return 0;
}

```

The output from runnig this is:
```
Loss: 0.693388
Loss: 0.693177
Loss: 0.693163
Loss: 0.693155
Loss: 0.693151
Loss: 0.693149
Loss: 0.693148
Loss: 0.693148
Loss: 0.693148
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
Loss: 0.693147
```

Please help. Best regards.

**Comment by OleStauning** [comment]
Ok. I think I figured this out. The non-convergence was caused by the ReLU dying... The solution was to use `LeakyReLU` instead. Now the XOR example converges.

**Comment by SandipGupta** [comment]
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CNTK;
using System.Collections;

namespace ConsoleApplication2
{
	class Program
	{
		static void Main(string[] args)
		{
			//double[,] data_in = new double[,] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } };
			//double[,] data_out = new double[,] { { 0 }, { 1 }, { 1 }, { 0 }};
			List<List<float>> data_in = new List<List<float>>();
			List<List<float>> data_out = new List<List<float>>();
			data_in.Add(new List<float> { 0, 0 });
			data_in.Add(new List<float> { 0, 1 });
			data_in.Add(new List<float> { 1, 0 });
			data_in.Add(new List<float> { 1, 1 });

			data_out.Add(new List<float> { 0 });
			data_out.Add(new List<float> { 1 });
			data_out.Add(new List<float> { 1 });
			data_out.Add(new List<float> { 0 });

			var device = DeviceDescriptor.CPUDevice;

			int inputDim = 2;
			int outputDim = 1;

			Variable input = Variable.InputVariable(new int[] { inputDim }, DataType.Float);
			Variable truth = Variable.InputVariable(new int[] { outputDim }, DataType.Float);
			Variable layer = input;
			NDShape shapeWeight = new NDShape(8, 2);
			NDShape shapeBias = new NDShape(8, 1);
			NDShape shapeSecondWeight = new NDShape(1, 8);
			NDShape shapeSecondBias = new NDShape(1);
			var hiddenWeights = new Parameter(NDArrayView.RandomUniform<float>(shapeWeight, -0.05, 0.05, 1, device));
			var hiddenBias = new Parameter(NDArrayView.RandomUniform<float>(shapeBias, -0.05, 0.05, 1, device));
			var hidden = CNTKLib.ReLU(CNTKLib.Plus(hiddenBias, CNTKLib.Times(hiddenWeights, input)));
			var outWeights = new Parameter(NDArrayView.RandomUniform<float>(shapeSecondWeight, -0.05, 0.05, 1, device));
			var outBias = new Parameter(NDArrayView.RandomUniform<float>(shapeSecondBias, -0.05, 0.05, 1, device));
			var prediction = CNTKLib.Sigmoid(CNTKLib.Plus(outBias, CNTKLib.Times(outWeights, hidden)));
			var trainingLoss = CNTKLib.BinaryCrossEntropy(prediction, truth);
			TrainingParameterScheduleDouble schedule = new CNTK.TrainingParameterScheduleDouble(
								0.05);
			var momentum = CNTKLib.MomentumAsTimeConstantSchedule(new DoubleVector { 1.0, 10.0, 100.0 });
			ParameterVector param = new ParameterVector();
			foreach(var item in prediction.Parameters())
			{
				param.Add(item);
			}
			var learner = CNTKLib.AdamLearner(param, schedule, momentum);
			LearnerVector learnerVector = new LearnerVector();
			learnerVector.Add(learner);
			var trainer = CNTKLib.CreateTrainer(prediction, trainingLoss, trainingLoss, learnerVector);

			const int minibatchSize = 10;

			List<List<float>> features = new List<List<float>>(minibatchSize);
			List<List<float>> labels = new List<List<float>>(minibatchSize);
			//std::vector<std::vector<float>> features(minibatchSize);
			//std::vector<std::vector<float>> labels(minibatchSize);

			int numMinibatchesToTrain = 200;
			int outputFrequencyInMinibatches = 10;
			int trainIdx = 0;
			for (int i = 0; i < numMinibatchesToTrain; ++i)
			{
				for (int j = 0; j < minibatchSize; ++j)
				{
					features[j] = data_in[trainIdx];
					labels[j] = data_out[trainIdx];
					trainIdx = (trainIdx + 1) % data_in.Count;
				}
				var batchInput = Value.CreateBatch(new int[] { inputDim }, features, device);
				var batchLabels = Value.CreateBatch(new int[] { outputDim }, labels, device);
				//IDictionary<Variable, Value> minibatchBindings = new Dictionary<Variable, Value>();
				//minibatchBindings.Add(input, batchInput);
				//minibatchBindings.Add(truth, batchLabels);

				//trainer.TrainMinibatch(minibatchBindings, device);
#pragma warning disable 618
				trainer.TrainMinibatch(
										new Dictionary<Variable, Value>() { { input, batchInput }, { truth, batchLabels } }, device);
#pragma warning restore 618
				if ((i % outputFrequencyInMinibatches) == 0)
					Console.WriteLine("Loss: ", trainer.PreviousMinibatchLossAverage());
			}
		}
	}
}



I am new in cntk,i have converted the same code in c# ..can u please tell me why it is not working ,if there are more errors please let me know,uploading correct solution in c# will be highly appreciated .Thank you so  much

**Comment by picolino** [comment]
Working XOR example on C#:

``` csharp
var inputs = new[] {new[] {0.0f, 0.0f}, new[] {1.0f, 0.0f}, new[] {0.0f, 1.0f}, new[] {1.0f, 1.0f}};
var expected = new[] {new[] {0.0f}, new[] {1.0f}, new[] {1.0f}, new[] {0.0f}};

var device = DeviceDescriptor.CPUDevice;

const int inputDimensions = 2;
const int outputDimensions = 1;

var inputVariable = Variable.InputVariable(new [] {inputDimensions}, DataType.Float);
var outputVariable = Variable.InputVariable(new [] {outputDimensions}, DataType.Float);

var hiddenWeights = new Parameter(NDArrayView.RandomUniform<float>(new [] {8 , 2}, -0.05, 0.05, 1, device));
var hiddenBias = new Parameter(NDArrayView.RandomUniform<float>(new [] { 8 }, -0.05, 0.05, 1, device));
var hidden = CNTKLib.LeakyReLU(CNTKLib.Plus(hiddenBias, CNTKLib.Times(hiddenWeights, inputVariable)), 0.1);

var outWeights = new Parameter(NDArrayView.RandomUniform<float>(new [] {1 , 8}, -0.05, 0.05, 1, device));
var outBias = new Parameter(NDArrayView.RandomUniform<float>(new [] { 1 }, -0.05, 0.05, 1, device));
var prediction = CNTKLib.Sigmoid(CNTKLib.Plus(outBias, CNTKLib.Times(outWeights, hidden)));
var trainingLoss = CNTKLib.BinaryCrossEntropy(prediction, outputVariable);

var schedule = new TrainingParameterScheduleDouble(0.05);
var momentum = CNTKLib.MomentumAsTimeConstantSchedule(new DoubleVector(new[] {1.0, 10.0, 100.0}));
var learner = CNTKLib.AdamLearner(new ParameterVector(prediction.Parameters().Select(o => o).ToList()), schedule, momentum);
var trainer = CNTKLib.CreateTrainer(prediction, trainingLoss, trainingLoss, new LearnerVector(new [] {learner}));

const int minibatchSize = 10;

var features = new float[minibatchSize][];
var labels = new float[minibatchSize][];

const int numMinibatchesToTrain = 4000;
const int outputFrequencyInMinibatches = 10;
var trainIndex = 0;

for (var i = 0; i < numMinibatchesToTrain; ++i)
{
    for(var j = 0; j < minibatchSize; ++j)
    {
        features[j] = inputs[trainIndex];
        labels[j] = expected[trainIndex];
        trainIndex = (trainIndex + 1) % inputs.Length;
    }
    var batchInput = Value.Create(new [] {inputDimensions}, features.Select(o => new NDArrayView(new [] {inputDimensions}, o, device)), device);
    var batchLabels = Value.Create(new [] {outputDimensions}, labels.Select(o => new NDArrayView(new [] {outputDimensions}, o, device)), device);
    var minibatchBindings = new Dictionary<Variable, Value>
                            {
                                {inputVariable, batchInput},
                                {outputVariable, batchLabels}
                            };

    trainer.TrainMinibatch(minibatchBindings, true, device);
    if (i % outputFrequencyInMinibatches == 0)
        System.Console.WriteLine($"Loss: {trainer.PreviousMinibatchLossAverage()}");
}
```

**Comment by adinelichleam** [comment]
I tweaked picolino's example to use TestHelper and remove NDArrayView so it can run on GPU.  Additionally, I harmonized it with the MNISTClassifier example to compare & contrast with the image classifier.

```
        public static void TrainAndEvaluate(DeviceDescriptor device)
        {
            // prepare training data
            LogicalMinibatchSource minibatchSource = new LogicalMinibatchSource(
                inputs: new[] { new[] { 0.0f, 0.0f }, new[] { 1.0f, 0.0f }, new[] { 0.0f, 1.0f }, new[] { 1.0f, 1.0f } },
                expected: new[] { new[] { 0.0f }, new[] { 1.0f }, new[] { 1.0f }, new[] { 0.0f } }
                );

            int inputDimensions = minibatchSource.InputDimensions;
            int hiddenDimensions = 4;
            int outputDimensions = minibatchSource.ExpectedDimensions;

            // build the network
            Variable inputLayer = Variable.InputVariable(new[] { inputDimensions }, DataType.Float);
            Function hiddenLayer = TestHelper.Dense(inputLayer, hiddenDimensions, device, Activation.LeakyReLU);
            Function outputLayer = TestHelper.Dense(hiddenLayer, outputDimensions, device, Activation.Sigmoid);

            Variable labels = Variable.InputVariable(new[] { outputDimensions }, DataType.Float);
            Function trainingLoss = CNTKLib.BinaryCrossEntropy(outputLayer, labels);
            Function prediction = trainingLoss;

            TrainingParameterScheduleDouble learningRatePerSample = new TrainingParameterScheduleDouble(0.25, 1);
            IList<Learner> parameterLearners = new List<Learner>() { Learner.SGDLearner(outputLayer.Parameters(), learningRatePerSample) };
            Trainer trainer = Trainer.CreateTrainer(outputLayer, trainingLoss, prediction, parameterLearners);

            const int minibatchSize = 10;

            const int numMinibatchesToTrain = 1000;
            const int outputFrequencyInMinibatches = 50;
            for (int i = 0; i < numMinibatchesToTrain; ++i)
            {
                LogicalMinibatchSource.Data minibatchData = minibatchSource.GetNextMinibatch(minibatchSize, device);
                Dictionary<Variable, Value> arguments = new Dictionary<Variable, Value>
                {
                    { inputLayer, minibatchData.Features },
                    { labels, minibatchData.Labels }
                };

                trainer.TrainMinibatch(arguments, true, device);
                TestHelper.PrintTrainingProgress(trainer, i, outputFrequencyInMinibatches);
            }
        }
    }
    public class LogicalMinibatchSource
    {
        public struct Data
        {
            public Value Features;
            public Value Labels;
        }

        private float[][] inputs;
        private float[][] expected;
        private int trainIndex = 0;

        public LogicalMinibatchSource(float[][] inputs, float[][] expected)
        {
            this.inputs = inputs;
            this.expected = expected;
        }

        public int InputDimensions { get { return inputs[0].Length; } }
        public int ExpectedDimensions { get { return expected[0].Length; } }

        public Data GetNextMinibatch(int minibatchSize, DeviceDescriptor device)
        {
            float[][] features = new float[minibatchSize][];
            float[][] labels = new float[minibatchSize][];
            for (int i = 0; i < minibatchSize; ++i)
            {
                features[i] = inputs[trainIndex];
                labels[i] = expected[trainIndex];
                trainIndex = (trainIndex + 1) % inputs.Length;
            }

            Data data;
            data.Features = Value.Create<float>(new[] { InputDimensions }, features, new bool[] { }, device);
            data.Labels = Value.Create<float>(new[] { ExpectedDimensions }, labels, new bool[] { }, device);
            return data;
        }
    }

```
